mq/plugins/remove_dots.py

```diff
@@ -0,0 +1,38 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at https://mozilla.org/MPL/2.0/.
+# Copyright (c) 2014 Mozilla Corporation
+
+
+class message(object):
+
+    def __init__(self):
+        '''
+        takes an incoming message
+        and checks for dots at the
+        start or end of the key and
+        removes them
+        '''
+
+        self.registration = ['cloudtrail']
+        self.priority = 5
+
+    def onMessage(self, message, metadata):
+        def renameKeysToRemoveDots(message):
+            if isinstance(message, dict):
+                message_keys = list(message.keys())
+                for key in message_keys:
+                    print(key)
+                    if key[0] == '.' or key[-1] == '.':
```

Comment:
I'd suggest using `str.startswith('.')` and `str.endswith('.')` here rather than string indexing, as it works safely with empty strings.

Reply:
Thank you for this, I was unaware of this ability!

Reply:
Doing it this way actually adds to the processing time this takes.

here's a small example:
In this test I'm working with the following initial output: 
output = {'key0': {'key1': 'value2', '.key2.': ['this', 'list'] , 'key3': {'.ds_store': 'value4', 'key5 with a .': 'value5'}}}

runtime using indices:
{'ds_store': 'value4', 'key5 with a ': 'value5'}
{'key1': 'value2', 'key3': None, 'key2': ['this', 'list']}
{'key0': None}
---- 0.00017595291137695312 seconds ----

runtime using startswith and endswith builtins:
{'ds_store': 'value4', 'key5 with a ': 'value5'}
{'key1': 'value2', 'key3': None, 'key2': ['this', 'list']}
{'key0': None}
---- 7.295608520507812e-05 seconds ----

I'll keep using the indices as otherwise we might start seeing this bottleneck.

Reply:
I've retested several times, and when doing a bazillion comparisons the indice is 2.5 times faster meaning the cycles we save in the interest of event processing is not worth the readability that startswith/endswith gives us.

Reply:
My concern is less with the readability and actually more the fault-tolerance. As I mentioned, if the length of the string happens to be 0 then using indices directly will result in a runtime exception.  Another way around this would be to check the length of the string ahead of time, writing something like

```py
if len(s) > 0 and (s[0] == '.' or s[-1] == '.'):
```

but that might have a similar or worse impact on the performance.

With that said I also don't think that squeezing fractions of performance is more important that writing more correct or readable code.  Not at this scale.

More to the point, though, it seems like using the startswith and endsiwth builtin is faster.  I guess this was an oversight because of the exponent?  0.00017 is 1.7e-04 and

```
>>> 7.295608520507812e-05 < 0.00017595291137695312
True
```

Reply:
When I made lower_keys I was asked to do testing on the time it takes before we could push that one to prod. That’s where this idea of performant came from.

here's more testing numbers for you so you can see exactly, I unfortunately had use two different methods for the numbers above:
phroztbyte:MozDef asmith$ python -m timeit -s 'text="foo"' 'text.startswith("a")'
2000000 loops, best of 5: 123 nsec per loop
phroztbyte:MozDef asmith$ python -m timeit -s 'text="foo"' 'text[0]=="a"'
5000000 loops, best of 5: 47.8 nsec per loop


We run every cloudtrail event through this, and at times we can have hundreds of thousands going through the pipe within minutes, this is why it can be critical

Reply:
I'll add the check for the key being null, but honestly since we are creating a list of keys that are present, there should never be an empty key, and if there was an empty key, it'd throw a parsing error way before this code is hit.